syntax = "proto3";

import "google/protobuf/wrappers.proto";
import "rpc_proto/rpc_types.proto";

package app;

enum AppEnum {
	EnumValue0 = 0;
	EnumValue1 = 1;
	EnumValue2 = 2;
	EnumValue3 = 3;
	EnumValue4 = 4;
}

message AppStruct {
	bool bool_value = 1;
	int32 int_value = 2;
	AppEnum enum_value = 3;
	string string_value = 4;
}

message AppEnumValue {
	option (nanorpc2.enum_wrapper) = true;
	AppEnum value = 1;
}

message MethodWithParameters_args {
	option (nanorpc2.expand_as_arguments) = true;
	string name = 1;
	int32 index = 2;
	AppStruct complext_data = 3;
}

service ITestApp {
//	rpc BoolValue(google.protobuf.BoolValue) returns (google.protobuf.BoolValue) {
//		option (nanorpc2.is_property) = true;
//	}

//	rpc IntValue(google.protobuf.Int32Value) returns (google.protobuf.Int32Value) {
//		option (nanorpc2.is_property) = true;
//	}

//	rpc EnumValue(AppEnumValue) returns (AppEnumValue) {
//		option (nanorpc2.is_property) = true;
//	}

//	rpc StringValue(google.protobuf.StringValue) returns (google.protobuf.StringValue)  {
//		option (nanorpc2.is_property) = true;
//	}

//	rpc StructValue(AppStruct) returns (AppStruct) {
//		option (nanorpc2.is_property) = true;
//	}

	rpc SyncVoidMethod(nanorpc2.RpcVoid) returns (nanorpc2.RpcVoid);
	rpc AsyncVoidMethod(nanorpc2.RpcVoid) returns (nanorpc2.RpcVoid) {
		option (nanorpc2.async) = true;
	}

	rpc SetStructReturnIntMethod(AppStruct) returns (google.protobuf.Int32Value);
	rpc GetStructMethod(google.protobuf.Int32Value) returns (AppStruct);

	rpc MethodWithoutParameters0(nanorpc2.RpcVoid) returns (nanorpc2.RpcVoid);
	rpc MethodWithoutParameters1(nanorpc2.RpcVoid) returns (google.protobuf.BoolValue);
	rpc MethodWithoutParameters2(nanorpc2.RpcVoid) returns (google.protobuf.StringValue);
	rpc MethodWithoutParameters3(nanorpc2.RpcVoid) returns (AppStruct);

	rpc MethodWithParameters0(MethodWithParameters_args) returns (nanorpc2.RpcVoid);
	rpc MethodWithParameters1(MethodWithParameters_args) returns (google.protobuf.BoolValue);
	rpc MethodWithParameters2(MethodWithParameters_args) returns (AppStruct);
}
